scripts/check-patch-apply.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/env python3
+# pylint: disable=invalid-name
+
+from argparse import ArgumentParser
+from pathlib import Path
+import os
+import subprocess
+import sys
+from tempfile import TemporaryDirectory
+
+parser = ArgumentParser(
+    description='Check that patches apply to their tree before running CI')
+parser.add_argument(
+    '-p',
+    '--patches-dir',
+    help='Path to patches directory (can be relative or absolute)',
+    required=True,
+    type=Path)
+parser.add_argument('-r',
+                    '--repo',
+                    help='URL to git repository',
+                    required=True)
+parser.add_argument('-R',
+                    '--ref',
+                    help='Git reference to apply patches upon',
+                    required=True)
+args = parser.parse_args()
+
+# If patches directory does not exist, there is nothing to check
+if not (patches_dir := args.patches_dir.resolve()).exists():
+    print(f"{patches_dir} does not exist, exiting 0...")
+    sys.exit(0)
+
+# There should be patches in there due to check-patches.py but we should double
+# check and fail if not
+if not list(patches_dir.glob('*.patch')):
+    print(f"{patches_dir} does not contain any patches?")
+    sys.exit(1)
+
+# Rather that invoke 'git clone', which can be expensive for servers depending
+# on the frequency and duration of requests, we fetch a tarball and 'git init'
+# that.
+with TemporaryDirectory() as workdir:
+    # Fetch the tarball from the repository. This is different for each type of
+    # tree that we support.
+    if args.repo.startswith(
```

Comment:
Looks like a good spot for Python's new-ish `match` expression. I'm always looking for an excuse to use it.

Not a blocker for approval though.